### ISSUE TITLE
Memory Issue with basic push/pull communication

### ISSUE THREAD

**Issue by ShaurCapella** [zeromq/pyzmq#1410]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description
Appear to have a memory leak where 2 TX processes send messages, 1 RX process receives, and memory usage increases continuously during transmission and doesn't go back down until I exit the program

# Environment
* libzmq version (commit hash if unreleased): oyzmq 19.01
* OS: Ubuntu 18.04

# Minimal test code / Steps to reproduce the issue
Run the below code, watch top or htop and see memory go up but never decrease, kill using ctrl-c, see memory released (also see printout of number of packets received equal 5000000*2 = 10000000 as expected 

```
import multiprocessing as mp
import time
import zmq 
ZMQ_FMT =  "tcp://127.0.0.1:%s"

def sender():
    context = zmq.Context()
    output = context.socket(zmq.PUSH)
    output.set_hwm(100000000)
    output.connect(ZMQ_FMT % 5999)

    a =  bytearray()
    count = 0
    size =2040
    num_packets = 100000000 // size
    for i in range(5000000):
        output.send( 1020 * b'10')
    print("DONE")
    while True:
        time.sleep(1)
            
def receiver():
    context = zmq.Context()
    input_zmq = context.socket(zmq.PULL)
    input_zmq.set_hwm(100000000)
    input_zmq.bind(ZMQ_FMT % 5999)
    count = 0
    try:
        while True:
        
            data = input_zmq.recv()
            count +=1
    except: 
        print (count)

mp.set_start_method('fork')
p1 = mp.Process(target=sender)
p2 = mp.Process(target=sender)
p3 = mp.Process(target=receiver)

p3.start()
p1.start()
p2.start()
while True:
    time.sleep(1)
```

# What's the actual result? (include assertion message & call stack if applicable)
HTOP shows memory increase but not any decrease after messages are consumed

# What's the expected result?
Memory usage shrinks back down after messages consumed

**Comment by minrk** [comment]
When I run this script (pyzmq-19.0.2, Python 3.7, mac), I do see memory of all processes returning to ~40MB when the messages actually complete delivery.

I believe the issue here is that messages haven't really finished sending when the last `send` has been called. The `send` method is async and you only really know that the actual network-send is complete when the socket has been closed *and* the context is terminated. If you add:

```python
print("DONE async sending")
output.close()
context.term()
print("DONE actually sending")
```

you'll see that it takes a pretty long time between the last `send` call and the last send to actually go over the wire (for me, it's about 20 seconds). Once the messages have actually left the memory of the sending process, they do indeed free their memory.

A message goes through four states before being freed:

1. Python send (hands off to queue in libzmq, still in sender memory)
2. low-level tcp send (transfer from sender memory to receiver memory, does not leave sender until complete receipt is confirmed)
3. in receiver libzmq queue (in receiver memory, no longer in sender memory)
4. Python recv (still in receiver memory)

all before it is finally freed and no longer resident anywhere.

I added some instrumentation to your script to log memory over time of each process, and this is the result:

![visualization](https://user-images.githubusercontent.com/151929/90246602-43c77700-de35-11ea-9017-6dfce5d30226.png)

In this chart, you can see that both senders finish their last send call around 62 seconds, but the receiver still has ~40% of the messages to receive at this point, indicating that a lot of messages are still waiting in libzmq, either in the receiver or the sender. From the memory profile, we can see it's backed up in the senders and they finally start emptying their backlog after the last send is requested. The ever-decreasing in the receiver after ~30 seconds suggests that the Python receiver *is* keeping up with the read-queue in libzmq, reading messages as fast as they arrive. On the other hand, the low-level send on the senders is *not* keeping up with the Python send calls, resulting in the send queues continuing to increase in size until getting very large, until they finally start emptying around 70 seconds and memory starts dropping. At the end, when all the messages have been actually received by the receiver, all processes are back to ~40MB resident.


This "libzmq can't keep up" situation is exactly the behavior governed by HWM. By setting HWM >= total messages, you are telling zmq that it should never stop filling memory with queued messages. HWM limits this queue size, and if HWM is hit (for PUSH/PULL sockets), the behavior is to block send until space becomes available. The result is that sends may take extra time, but it also means that the send queue has a limited size in memory.

Setting HWM to 10000 on both sides results in this stable memory profile (note: it's also ~25% faster!):

![visualization-2](https://user-images.githubusercontent.com/151929/90248933-e08c1380-de39-11ea-9234-3c67f296bdcb.png)

and it looks much the same, scaled up, with HWM=1_000_000:

![visualization-3](https://user-images.githubusercontent.com/151929/90249010-031e2c80-de3a-11ea-9597-1b970889edef.png)


In short, if you are concerned about memory growth, HWM lets you put a limit on this in the case of libzmq/tcp not keeping up.

**Comment by roblight** [comment]
@minrk did you run your test under Linux?  We're having the same problem but only under Linux, not macOS.  Thanks!
